### PR TITLE
Update monitoring job to query timestamp metadata from db instead of v2 feed

### PR DESCRIPTION
**Pull request by xavierdecoster**

### CODE REVIEW

File: src/Catalog/Helpers/GalleryDatabaseQueryService.cs [hunk 1]
```diff
@@ -48,6 +51,44 @@ public class GalleryDatabaseQueryService : IGalleryDatabaseQueryService
                 Db2CatalogCursor.ByLastEdited(since, top));
         }
 
+        public async Task<FeedPackageDetails> GetPackage(string id, string version)
```
Comment (loic-sharma) [inline src/Catalog/Helpers/GalleryDatabaseQueryService.cs]: >GetPackage [](start = 46, length = 10)

`GetPackageDetailsAsync` records a duration metric for each query it executes, consider adding that to `GetPackage` as well

File: src/Catalog/Helpers/IGalleryDatabaseQueryService.cs [hunk 2]
```diff
@@ -9,6 +9,7 @@ namespace NuGet.Services.Metadata.Catalog.Helpers
 {
     public interface IGalleryDatabaseQueryService
     {
+        Task<FeedPackageDetails> GetPackage(string id, string version);
```
Comment (loic-sharma) [inline src/Catalog/Helpers/IGalleryDatabaseQueryService.cs]: >GetPackage [](start = 33, length = 10)

What do you think of the name `GetPackageOrNull` to clarify the 'not found' behavior?

File: src/NuGet.Services.Metadata.Catalog.Monitoring/Providers/PackageTimestampMetadataResourceV2Provider.cs [hunk 3]
```diff
@@ -7,15 +7,21 @@
...
 using NuGet.Protocol;
 using NuGet.Protocol.Core.Types;
+using NuGet.Services.Metadata.Catalog.Helpers;
 
 namespace NuGet.Services.Metadata.Catalog.Monitoring
 {
     public class PackageTimestampMetadataResourceV2Provider : ResourceProvider
     {
-        public PackageTimestampMetadataResourceV2Provider(ILoggerFactory loggerFactory)
+        private readonly IGalleryDatabaseQueryService _galleryDatabaseQueryService;
```
Comment (loic-sharma) [inline src/NuGet.Services.Metadata.Catalog.Monitoring/Providers/PackageTimestampMetadataResourceV2Provider.cs]: >_galleryDatabaseQueryService [](start = 54, length = 28)

This name is a little verbose. What do you think of a name like `_galleryDatabase`? If you change this name, please also update `PackageTimestampMetadataResource`.

File: src/Catalog/Helpers/GalleryDatabaseQueryService.cs (hunk 1 above)
Comment (xavierdecoster) [reply src/Catalog/Helpers/GalleryDatabaseQueryService.cs]: Added in https://github.com/NuGet/NuGet.Services.Metadata/pull/537/commits/ff3e15d46eda897db0d1ab14ce4a608cc31794c0

File: src/Catalog/Helpers/IGalleryDatabaseQueryService.cs (hunk 2 above)
Comment (xavierdecoster) [reply src/Catalog/Helpers/IGalleryDatabaseQueryService.cs]: Better! Fixed in https://github.com/NuGet/NuGet.Services.Metadata/pull/537/commits/6e19b926fed8908b24abe4f0597486073991c65c

File: src/NuGet.Services.Metadata.Catalog.Monitoring/Providers/PackageTimestampMetadataResourceV2Provider.cs (hunk 3 above)
Comment (xavierdecoster) [reply src/NuGet.Services.Metadata.Catalog.Monitoring/Providers/PackageTimestampMetadataResourceV2Provider.cs]: Addressed with https://github.com/NuGet/NuGet.Services.Metadata/pull/537/commits/efccb258c8c58307490c364713a284d0fa12322e

File: src/NuGet.Services.Metadata.Catalog.Monitoring/Model/PackageTimestampMetadata.cs [hunk 4]
```diff
@@ -26,9 +26,9 @@ public class PackageTimestampMetadata : INuGetResource
         public DateTime? Last => new[] { Created, LastEdited, Deleted }.Max();
 
         /// <summary>
-        /// Creates a <see cref="PackageTimestampMetadata"/> that represents a package that exists on the feed.
+        /// Creates a <see cref="PackageTimestampMetadata"/> that represents a package that exists on the package repository.
```
[the remark this answers is not in the document]
Comment (xavierdecoster) [reply src/NuGet.Services.Metadata.Catalog.Monitoring/Model/PackageTimestampMetadata.cs]: Fixed in https://github.com/NuGet/NuGet.Services.Metadata/pull/537/commits/1c2ca1e94d5607d83d6fd10ada0aa5f5ecd8cdf0

File: src/Catalog/Helpers/GalleryDatabaseQueryService.cs [hunk 5]
```diff
@@ -48,6 +51,47 @@ public class GalleryDatabaseQueryService : IGalleryDatabaseQueryService
...
+                            while (await packagesReader.ReadAsync())
+                            {
+                                packages.Add(_db2catalogProjection.FromDataRecord(packagesReader));
+                            }
+                        }
+                    }
+                }
+            }
+
+            return packages.FirstOrDefault();
```
[the remark this answers is not in the document]
Comment (xavierdecoster) [reply src/Catalog/Helpers/GalleryDatabaseQueryService.cs]: Changed `FirstOrDefault` into `SingleOrDefault` which will throw if more than one element is returned. https://github.com/NuGet/NuGet.Services.Metadata/pull/537/commits/a61d3959fffb14990a415847d305ebcda761eb49

File: src/Catalog/Helpers/GalleryDatabaseQueryService.cs [hunk 6]
```diff
@@ -126,6 +170,29 @@ internal static string BuildDb2CatalogSqlQuery(Db2CatalogCursor cursor)
                     ORDER BY P.[{cursor.ColumnName}]";
         }
 
+        /// <summary>
+        /// Builds the parameterized SQL query for retrieving package details given an id and version.
+        /// </summary>
+        internal static string BuildGetPackageSqlQuery()
+        {
+            return $@"SELECT 
+                        PR.[Id],
```
[the remark this answers is not in the document]
Comment (xavierdecoster) [reply src/Catalog/Helpers/GalleryDatabaseQueryService.cs]: Good idea! Fixed in https://github.com/NuGet/NuGet.Services.Metadata/pull/537/commits/95073da766f8e7a747372973d4892024f00d6ddf

File: src/Catalog/Helpers/GalleryDatabaseQueryService.cs [hunk 7]
```diff
@@ -126,6 +170,29 @@ internal static string BuildDb2CatalogSqlQuery(Db2CatalogCursor cursor)
...
+                        P.[LastEdited],
+                        P.[Published],
+                        P.[Listed],
+                        P.[HideLicenseReport],
+                        P.[LicenseNames],
+                        P.[LicenseReportUrl]
+                    FROM [dbo].[Packages] AS P
+                    INNER JOIN [dbo].[PackageRegistrations] AS PR ON P.[PackageRegistrationKey] = PR.[Key]
+                    WHERE P.[PackageStatusKey] = {(int)PackageStatus.Available} 
+                        AND P.[Deleted] = 0
```
[the remark this answers is not in the document]
Comment (xavierdecoster) [reply src/Catalog/Helpers/GalleryDatabaseQueryService.cs]: Hmm... true! Perhaps we should get rid of the `Deleted` column completely.
Addressed with https://github.com/NuGet/NuGet.Services.Metadata/pull/537/commits/6bdfb478014c9d41ebf4dde0817a9c55991db30e

**Pull request merged**